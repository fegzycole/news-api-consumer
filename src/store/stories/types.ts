export const SET_STORIES = "SET_STORIES";
export const SET_ERROR = "SET_ERROR";
export const SET_FETCHING_STORIES = "SET_FETCHING_STORIES";

export enum StoryCategories {
  TopHeadlines = "Top Headlines",
  Business = "Business",
  Entertainment = "Entertainment",
  General = "General",
  Health = "Health",
  Science = "Science",
  Sports = "Sports",
  Technology = "Technology",
}

export interface StoryData {
  source: {
    id: null | number;
    name: string;
  };
  author: string;
  title: string;
  description: string;
  url: string;
  urlToImage: string;
  publishedAt: Date | string;
  content: string;
}

export interface StoriesState {
  stories: StoryData[];
  fetchingStories: boolean;
  error?: string;
}

export interface AddStories {
  stories: StoryData[];
  clearStories: boolean;
}

export interface SetStories {
  type: typeof SET_STORIES;
  payload: AddStories;
}

export interface SetError {
  type: typeof SET_ERROR;
  payload: string;
}

export interface SetFetchingStories {
  type: typeof SET_FETCHING_STORIES;
  payload: boolean;
}

export type MoviesActionTypes =
  | SetStories
  | SetError
  | SetFetchingStories;
