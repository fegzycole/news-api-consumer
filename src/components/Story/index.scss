@import "../../styles/mixins";

@mixin standard_story_text($letter-spacing, $color) {
  @include text_styles(700, 11px, $color, 'normal', $letter-spacing);
  @include standard-margin-bottom;
  @include transform-uppercase;
}

@mixin get_time_styles($font-size) {
  @include text-styles(300, $font-size, $light-black-two, 18px);
  @include standard-margin-bottom(15px);
  opacity: 0.78;
}

@mixin get_title_styles($font-size, $line-height) {
  @include standard-margin-bottom;
  @include text-styles(700, $font-size, $primary-color, $line-height);
}

.Story_Header {
  @include standard_story_text(0.3em, $primary-color);
  color: $primary-color;
}

.Story_Image {
  width: 100%;
  @include standard-margin-bottom;
}

.Story_Source {
  @include standard_story_text(0.15em, $secondary-color);
}

.Top_Story_Title {
  @include get_title_styles(24px, 30px);
}

.Story_Title {
  @include get_title_styles(18px, 22px);
}

.Top_Story_Description {
  @include text_styles(400, 15px, $light-black-one, 21px);
  @include standard-margin-bottom;
}

.Story_Description {
  display: none;
}

.Top_Story_Time {
  @include get_time_styles(12px);
}

.Story_Time {
  @include get_time_styles(11px);
}

.Story_Util {
  @include standard-margin-bottom(15px);
}

.Story_Section {
  @include lgScreen {
    @include flex-styles(row, space-between, center);
    width: 100%;
  }
}

.Story_Section_Item {
  @include lgScreen {
    width: 48%;
  }
}

.Story_Section_Centered {
  @include lgScreen {
    @include flex-styles(column, center);
  }
}

.Reverse {
  flex-direction: row-reverse;
}
