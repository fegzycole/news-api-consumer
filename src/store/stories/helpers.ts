import { StoryCategories } from "./types";

export const apiKey = process.env.REACT_APP_API_KEY;
export const apiUrl = "https://newsapi.org/v2/top-headlines";

export const getRequestPath = (
  category: StoryCategories,
  pageNumber: number
): string => {
  let path = `${apiUrl}?apiKey=${apiKey}&pageSize=10&page=${pageNumber}&language=en&sortBy=publishedAt`;

  if (category !== StoryCategories.TopHeadlines) {
    path = `${path}&category=${category.toLowerCase()}`;
  }

  return path;
};

export const getResponseError = (error: any): string => {
  let serverErrResponse = error?.response?.data?.message;

  if (!serverErrResponse) {
    serverErrResponse = error.message;
  }

  return serverErrResponse;
};
