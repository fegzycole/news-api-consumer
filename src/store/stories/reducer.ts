import {
  MoviesActionTypes,
  StoriesState,
  SET_STORIES,
  SET_ERROR,
  SET_FETCHING_STORIES,
} from "./types";

const initialState: StoriesState = {
  stories: [],
  fetchingStories: false,
  error: undefined,
};

const storiesReducer = (
  state: StoriesState = initialState,
  action: MoviesActionTypes
) => {
  switch (action.type) {
    case SET_STORIES:
      return {
        ...state,
        stories: action.payload.clearStories
          ? action.payload.stories
          : state.stories.concat(action.payload.stories),
        fetchingStories: false,
      };
    case SET_ERROR:
      return {
        ...state,
        error: action.payload,
        fetchingStories: false,
      };
    case SET_FETCHING_STORIES:
      return {
        ...state,
        fetchingStories: action.payload,
      };
    default:
      return state;
  }
};

export default storiesReducer;
