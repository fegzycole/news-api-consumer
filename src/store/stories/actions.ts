import { Dispatch } from "react";
import axios from "axios";

import {
  StoryData,
  SetStories,
  SET_STORIES,
  SetError,
  SET_ERROR,
  SetFetchingStories,
  SET_FETCHING_STORIES,
  StoryCategories,
  AddStories,
} from "./types";
import { getResponseError, getRequestPath } from "./helpers";

export const setStories = (payload: AddStories): SetStories => {
  return {
    type: SET_STORIES,
    payload,
  };
};

export const setError = (payload: string): SetError => {
  return {
    type: SET_ERROR,
    payload,
  };
};

export const setFetchingStories = (payload: boolean): SetFetchingStories => {
  return {
    type: SET_FETCHING_STORIES,
    payload,
  };
};

export const getStories = (
  category: StoryCategories,
  pageNumber: number
): ((
  dispatch: Dispatch<{ type: string; payload?: unknown }>
) => Promise<void>) => {
  return async (dispatch) => {
    dispatch(setError(""));
    dispatch(setFetchingStories(true));

    const requestPath = getRequestPath(category, pageNumber);

    try {
      const request = await axios.get(requestPath);

      const responseData = request.data.articles as StoryData[];

      const stories: StoryData[] = responseData as StoryData[];

      const payload: AddStories = {
        stories,
        clearStories: pageNumber === 1,
      };

      return dispatch(setStories(payload));
    } catch (error: any) {
      const renderedError = getResponseError(error);
      return dispatch(setError(renderedError));
    }
  };
};
