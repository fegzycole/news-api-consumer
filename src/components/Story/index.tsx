import moment from "moment";

import { StoryCategories, StoryData } from "../../store/stories/types";
import Divider from "../Divider";
import DefaultIcon from '../../assets/default_news.jpeg';
import "./index.scss";

export interface TopStoryProps {
  story: StoryData;
  selectedCategory: StoryCategories;
  topStory?: boolean;
  reverse?: boolean;
}

const TopStory: React.FC<TopStoryProps> = ({
  story,
  selectedCategory,
  topStory,
  reverse,
}) => {
  let storyHeader, storyTime, classPrefix;

  if (topStory) {
    if (selectedCategory === StoryCategories.TopHeadlines) {
      storyHeader = "breaking news";
    } else {
      storyHeader = `latest ${selectedCategory} news`;
    }

    storyTime = moment(story.publishedAt).fromNow();
    classPrefix = "Top_Story";
  } else {
    storyTime = `${story.author} - ${moment(story.publishedAt).format("LL")}`;
    classPrefix = "Story";
  }

  return (
    <article className="Story">
      <a href={story.url}>
        <h5 className="Story_Header">{storyHeader}</h5>
        <section className={`${classPrefix}_Section ${reverse && "Reverse"}`}>
          <div className={`${classPrefix}_Section_Item`}>
            <img
              src={story.urlToImage || DefaultIcon}
              alt={story.title}
              className="Story_Image"
            />
          </div>
          <div
            className={`${classPrefix}_Section_Item ${classPrefix}_Section_Centered`}
          >
            <h6 className="Story_Source">{story.source.name}</h6>
            <h3 className={`${classPrefix}_Title`}>{story.title}</h3>
            <h4 className={`${classPrefix}_Description`}>
              {story.description}
            </h4>
            <p className={`${classPrefix}_Time`}>{storyTime}</p>
          </div>
        </section>
        <Divider />
        <div className="Story_Util" />
      </a>
    </article>
  );
};

export default TopStory;
