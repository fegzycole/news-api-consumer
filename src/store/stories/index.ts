import * as Types from './types';
import storiesReducer from './reducer';
import * as Actions from './actions';

export { Types, storiesReducer, Actions };
