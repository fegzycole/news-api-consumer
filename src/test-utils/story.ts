import { StoryData } from "../store/stories/types";

const generateStories = () => {
  const mockStories: StoryData[] = [];

  let i = 0;

  while (i < 12) {
    mockStories.push(mockStory);

    i++;
  }

  return mockStories;
};

export const mockStory: StoryData = {
  publishedAt: "Mock Date",
  author: "Mock Author",
  urlToImage: "MockUrlToImage",
  description: "Mock Description",
  source: {
    name: "Mock Source Name",
    id: 1,
  },
  title: "Mock Title",
  url: "mockUrl",
  content: "MockContent",
};

export const mockStories: StoryData[] = generateStories();
