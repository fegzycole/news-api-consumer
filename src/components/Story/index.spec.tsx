import { render } from "@testing-library/react";
import { StoryCategories } from "../../store/stories/types";
import { mockStory } from "../../test-utils/story";
import Story, { TopStoryProps } from "./index";

const mockProps: TopStoryProps = {
  story: mockStory,
  selectedCategory: StoryCategories.Science,
};

jest.mock("moment", () => () => ({
  format: jest.fn().mockImplementation(() => "Aug 21 2022"),
  fromNow: jest.fn().mockImplementation(() => "1 min ago"),
}));

jest.mock("../Divider", () => ({
  __esModule: true,
  default: () => {
    return <div className="Divider" />;
  },
  Divider: true,
}));

describe("<Story />", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should match the Story component snapshot", () => {
    const view = render(<Story {...mockProps} />);
    expect(view.container).toMatchSnapshot();
  });

  describe("when the story is a top story", () => {
    it("should match the Story component snapshot for when it is a top story", () => {
      const view = render(<Story {...mockProps} topStory />);
      expect(view.container).toMatchSnapshot();
    });
  });

  describe("when the flex orientation should be reversed", () => {
    it("should match the Story component snapshot for when the flex orientation should be reversed", () => {
      const view = render(<Story {...mockProps} reverse />);
      expect(view.container).toMatchSnapshot();
    });
  });
});
