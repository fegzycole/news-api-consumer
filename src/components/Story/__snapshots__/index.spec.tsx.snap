// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Story /> should match the Story component snapshot 1`] = `
<div>
  <article
    class="Story"
  >
    <a
      href="mockUrl"
    >
      <h5
        class="Story_Header"
      />
      <section
        class="Story_Section undefined"
      >
        <div
          class="Story_Section_Item"
        >
          <img
            alt="Mock Title"
            class="Story_Image"
            src="MockUrlToImage"
          />
        </div>
        <div
          class="Story_Section_Item Story_Section_Centered"
        >
          <h6
            class="Story_Source"
          >
            Mock Source Name
          </h6>
          <h3
            class="Story_Title"
          >
            Mock Title
          </h3>
          <h4
            class="Story_Description"
          >
            Mock Description
          </h4>
          <p
            class="Story_Time"
          >
            Mock Author - Aug 21 2022
          </p>
        </div>
      </section>
      <div
        class="Divider"
      />
      <div
        class="Story_Util"
      />
    </a>
  </article>
</div>
`;

exports[`<Story /> when the flex orientation should be reversed should match the Story component snapshot for when the flex orientation should be reversed 1`] = `
<div>
  <article
    class="Story"
  >
    <a
      href="mockUrl"
    >
      <h5
        class="Story_Header"
      />
      <section
        class="Story_Section Reverse"
      >
        <div
          class="Story_Section_Item"
        >
          <img
            alt="Mock Title"
            class="Story_Image"
            src="MockUrlToImage"
          />
        </div>
        <div
          class="Story_Section_Item Story_Section_Centered"
        >
          <h6
            class="Story_Source"
          >
            Mock Source Name
          </h6>
          <h3
            class="Story_Title"
          >
            Mock Title
          </h3>
          <h4
            class="Story_Description"
          >
            Mock Description
          </h4>
          <p
            class="Story_Time"
          >
            Mock Author - Aug 21 2022
          </p>
        </div>
      </section>
      <div
        class="Divider"
      />
      <div
        class="Story_Util"
      />
    </a>
  </article>
</div>
`;

exports[`<Story /> when the story is a top story should match the Story component snapshot for when it is a top story 1`] = `
<div>
  <article
    class="Story"
  >
    <a
      href="mockUrl"
    >
      <h5
        class="Story_Header"
      >
        latest Science news
      </h5>
      <section
        class="Top_Story_Section undefined"
      >
        <div
          class="Top_Story_Section_Item"
        >
          <img
            alt="Mock Title"
            class="Story_Image"
            src="MockUrlToImage"
          />
        </div>
        <div
          class="Top_Story_Section_Item Top_Story_Section_Centered"
        >
          <h6
            class="Story_Source"
          >
            Mock Source Name
          </h6>
          <h3
            class="Top_Story_Title"
          >
            Mock Title
          </h3>
          <h4
            class="Top_Story_Description"
          >
            Mock Description
          </h4>
          <p
            class="Top_Story_Time"
          >
            1 min ago
          </p>
        </div>
      </section>
      <div
        class="Divider"
      />
      <div
        class="Story_Util"
      />
    </a>
  </article>
</div>
`;
